File: objectified-ui/lib/repositories/providers/github-provider.ts
import { RepositoryProviderError, type RepositoryProvider } from './repository-provider';
import type { BranchInfo, ListReposOpts, RepoDetail, RepoRef, RepoSummary, TreeEntry } from './types';

type FetchFn = typeof fetch;

interface RequestJsonOptions {
  path: string;
  token: string;
  query?: Record<string, string | number | undefined>;
}

const GITHUB_MAX_PER_PAGE = 100;

export class GithubRepositoryProvider implements RepositoryProvider {
  readonly id = 'github' as const;
  private readonly fetchFn: FetchFn;

  constructor(fetchFn?: FetchFn) {
    this.fetchFn = fetchFn ?? fetch;
  }

  async *listRepositories(token: string, opts?: ListReposOpts): AsyncIterable<RepoSummary> {
    const perPage = Math.min(opts?.perPage ?? GITHUB_MAX_PER_PAGE, GITHUB_MAX_PER_PAGE);
    let page = opts?.page ?? 1;
    const sort = opts?.sort ?? 'updated';

    while (true) {
      const repositories = await this.requestJson<unknown[]>({
        path: '/user/repos',
        token,
        query: {
          per_page: perPage,
          page,
          sort,
          visibility: opts?.visibility,
        },
      });

      if (!Array.isArray(repositories) || repositories.length === 0) {
        return;
      }

      for (const entry of repositories) {
        if (!entry || typeof entry !== 'object') {
          continue;
        }

        const repo = entry as Record<string, unknown>;
        yield {
          id: String(repo.id ?? ''),
          name: String(repo.name ?? ''),
          fullName: String(repo.full_name ?? ''),
          description: typeof repo.description === 'string' ? repo.description : null,
          isPrivate: Boolean(repo.private),
          defaultBranch: typeof repo.default_branch === 'string' ? repo.default_branch : 'main',
          htmlUrl: String(repo.html_url ?? ''),
          updatedAt: typeof repo.updated_at === 'string' ? repo.updated_at : null,
        };
      }

      if (repositories.length < perPage) {
        return;
      }

      page += 1;
    }
  }

  async getRepository(token: string, owner: string, name: string): Promise<RepoDetail> {
    const repo = await this.requestJson<Record<string, unknown>>({
      path: `/repos/${encodeURIComponent(owner)}/${encodeURIComponent(name)}`,
      token,
    });

    return {
      id: String(repo.id ?? ''),
      name: String(repo.name ?? ''),
      fullName: String(repo.full_name ?? `${owner}/${name}`),
      description: typeof repo.description === 'string' ? repo.description : null,
      isPrivate: Boolean(repo.private),
      defaultBranch: typeof repo.default_branch === 'string' ? repo.default_branch : 'main',
      htmlUrl: String(repo.html_url ?? ''),
      updatedAt: typeof repo.updated_at === 'string' ? repo.updated_at : null,
    };
  }

  async *listBranches(token: string, repo: RepoRef): AsyncIterable<BranchInfo> {
    let page = 1;
    const perPage = 100;

    while (true) {
      const response = await this.requestJson<unknown[]>({
        path: `/repos/${encodeURIComponent(repo.owner)}/${encodeURIComponent(repo.name)}/branches`,
        token,
        query: { per_page: perPage, page },
      });

      if (!Array.isArray(response) || response.length === 0) {
        return;
      }

      for (const branch of response) {
        if (!branch || typeof branch !== 'object') {
          continue;
        }

        const raw = branch as Record<string, unknown>;
        const commit = (raw.commit as Record<string, unknown> | undefined) ?? {};
        yield {
          name: String(raw.name ?? ''),
          commitSha: String(commit.sha ?? ''),
          isProtected: Boolean(raw.protected),
        };
      }

      if (response.length < perPage) {
        return;
      }

      page += 1;
    }
  }

  async getCommitSha(token: string, repo: RepoRef, branch: string): Promise<string> {
    const commit = await this.requestJson<Record<string, unknown>>({
      path: `/repos/${encodeURIComponent(repo.owner)}/${encodeURIComponent(repo.name)}/commits/${encodeURIComponent(branch)}`,
      token,
    });
    return String(commit.sha ?? '');
  }

  async *walkTree(token: string, repo: RepoRef, branch: string, subpath?: string): AsyncIterable<TreeEntry> {
    const normalizedSubpath = normalizeSubpath(subpath);
    const response = await this.requestJson<Record<string, unknown>>({
      path: `/repos/${encodeURIComponent(repo.owner)}/${encodeURIComponent(repo.name)}/git/trees/${encodeURIComponent(branch)}`,
      token,
      query: { recursive: 1 },
    });
    const tree = response.tree;

    if (!Array.isArray(tree)) {
      return;
    }

    for (const item of tree) {
      if (!item || typeof item !== 'object') {
        continue;
      }

      const raw = item as Record<string, unknown>;
      const rawPath = String(raw.path ?? '');
      if (!rawPath) {
        continue;
      }

      if (normalizedSubpath && !rawPath.startsWith(normalizedSubpath)) {
        continue;
      }

      const type = mapTreeType(raw.type);
      if (!type) {
        continue;
      }

      yield {
        path: rawPath,
        type,
        sha: String(raw.sha ?? ''),
        sizeBytes: Number(raw.size ?? 0),
      };
    }
  }

  private encodeContentPath(path: string): string {
    return path
      .replace(/^\/+|\/+$/g, '')
      .split('/')
      .filter((segment) => segment.length > 0)
      .map((segment) => encodeURIComponent(segment))
      .join('/');
  }

  async readFile(
    token: string,
    repo: RepoRef,
    branch: string,
    path: string
  ): Promise<{ contentBase64: string; sha: string; sizeBytes: number }> {
    const encodedPath = this.encodeContentPath(path);
    const content = await this.requestJson<Record<string, unknown>>({
      path: `/repos/${encodeURIComponent(repo.owner)}/${encodeURIComponent(repo.name)}/contents/${encodedPath}`,
      token,
      query: { ref: branch },
    });

    const rawContent = typeof content.content === 'string' ? content.content : '';
    return {
      contentBase64: rawContent.replace(/\n/g, ''),
      sha: String(content.sha ?? ''),
      sizeBytes: Number(content.size ?? 0),
    };
  }

  private async requestJson<T>({ path, token, query }: RequestJsonOptions): Promise<T> {
    const searchParams = new URLSearchParams();
    for (const [key, value] of Object.entries(query ?? {})) {
      if (value !== undefined) {
        searchParams.set(key, String(value));
      }
    }
    const suffix = searchParams.size > 0 ? `?${searchParams.toString()}` : '';
    const url = `https://api.github.com${path}${suffix}`;

    let response: Response;
    try {
      response = await this.fetchFn(url, {
        headers: {
          Authorization: `Bearer ${token}`,
          Accept: 'application/vnd.github+json',
          'X-GitHub-Api-Version': '2022-11-28',
        },
      });
    } catch (error) {
      throw new RepositoryProviderError('NETWORK', 'Failed to reach GitHub API', undefined, { cause: error });
    }

    if (!response.ok) {
      const body = await response.text();
      throw normalizeError(response.status, body, response.headers);
    }

    return (await response.json()) as T;
  }
}

function normalizeSubpath(subpath?: string): string {
  if (!subpath) {
    return '';
  }
  const trimmed = subpath.replace(/^\/+/, '').replace(/\/+$/, '');
  if (!trimmed) {
    return '';
  }
  return `${trimmed}/`;
}

function mapTreeType(value: unknown): TreeEntry['type'] | null {
  switch (value) {
    case 'blob':
      return 'file';
    case 'tree':
      return 'dir';
    case 'symlink':
      return 'symlink';
    case 'commit':
      return 'submodule';
    default:
      return null;
  }
}

function normalizeError(status: number, body: string, headers: Headers): RepositoryProviderError {
  if (status === 401) {
    return new RepositoryProviderError('UNAUTHORIZED', 'Unauthorized GitHub token', status);
  }
  if (status === 404) {
    return new RepositoryProviderError('NOT_FOUND', 'GitHub resource not found', status);
  }
  if (status === 429 || isRateLimited(status, body, headers)) {
    return new RepositoryProviderError('RATE_LIMITED', 'GitHub API rate limit exceeded', status);
  }
  return new RepositoryProviderError('UNKNOWN', 'Unexpected GitHub API error', status);
}

function isRateLimited(status: number, body: string, headers: Headers): boolean {
  if (status !== 403) {
    return false;
  }

  const remaining = headers.get('x-ratelimit-remaining');
  if (remaining === '0') {
    return true;
  }

  return body.toLowerCase().includes('rate limit');
}

File: objectified-rest/src/app/repositories/providers/github_provider.py
from __future__ import annotations

from typing import Any, AsyncIterator
from urllib.parse import quote

import httpx

from .base import (
    BranchInfo,
    ListReposOpts,
    ReadFileResult,
    RepoDetail,
    RepoRef,
    RepositoryProvider,
    RepositoryProviderError,
    RepositoryProviderErrorCode,
    RepositoryProviderId,
    RepoSummary,
    TreeEntry,
)


class GithubRepositoryProvider(RepositoryProvider):
    id = RepositoryProviderId.GITHUB

    def __init__(self, client: httpx.AsyncClient | None = None) -> None:
        self._external_client = client
        self._owned_client: httpx.AsyncClient | None = None

    def _get_or_create_client(self) -> httpx.AsyncClient:
        if self._external_client is not None:
            return self._external_client
        if self._owned_client is None:
            self._owned_client = httpx.AsyncClient(base_url="https://api.github.com", timeout=20)
        return self._owned_client

    async def aclose(self) -> None:
        if self._owned_client is not None:
            await self._owned_client.aclose()
            self._owned_client = None

    async def __aenter__(self) -> "GithubRepositoryProvider":
        return self

    async def __aexit__(self, *args: object) -> None:
        await self.aclose()

    async def list_repositories(self, token: str, opts: ListReposOpts | None = None) -> AsyncIterator[RepoSummary]:
        current = opts or ListReposOpts()
        page = current.page
        per_page = min(current.per_page, 100)

        while True:
            repositories = await self._request_json(
                f"/user/repos?sort={current.sort}&per_page={per_page}&page={page}&visibility={current.visibility}",
                token,
            )
            if not isinstance(repositories, list) or len(repositories) == 0:
                return

            for repo in repositories:
                if not isinstance(repo, dict):
                    continue
                yield RepoSummary(
                    id=str(repo.get("id", "")),
                    name=str(repo.get("name", "")),
                    full_name=str(repo.get("full_name", "")),
                    description=repo.get("description"),
                    is_private=bool(repo.get("private", False)),
                    default_branch=str(repo.get("default_branch", "main")),
                    html_url=str(repo.get("html_url", "")),
                    updated_at=repo.get("updated_at"),
                )

            if len(repositories) < per_page:
                return
            page += 1

    async def get_repository(self, token: str, owner: str, name: str) -> RepoDetail:
        repo = await self._request_json(f"/repos/{owner}/{name}", token)
        if not isinstance(repo, dict):
            raise RepositoryProviderError(RepositoryProviderErrorCode.UNKNOWN, "Malformed GitHub response")

        return RepoDetail(
            id=str(repo.get("id", "")),
            name=str(repo.get("name", "")),
            full_name=str(repo.get("full_name", f"{owner}/{name}")),
            description=repo.get("description"),
            is_private=bool(repo.get("private", False)),
            default_branch=str(repo.get("default_branch", "main")),
            html_url=str(repo.get("html_url", "")),
            updated_at=repo.get("updated_at"),
        )

    async def list_branches(self, token: str, repo: RepoRef) -> AsyncIterator[BranchInfo]:
        page = 1
        per_page = 100

        while True:
            branches = await self._request_json(
                f"/repos/{repo.owner}/{repo.name}/branches?per_page={per_page}&page={page}",
                token,
            )
            if not isinstance(branches, list) or len(branches) == 0:
                return

            for branch in branches:
                if not isinstance(branch, dict):
                    continue
                commit = branch.get("commit", {})
                if not isinstance(commit, dict):
                    commit = {}
                yield BranchInfo(
                    name=str(branch.get("name", "")),
                    commit_sha=str(commit.get("sha", "")),
                    is_protected=bool(branch.get("protected", False)),
                )

            if len(branches) < per_page:
                return
            page += 1

    async def get_commit_sha(self, token: str, repo: RepoRef, branch: str) -> str:
        commit = await self._request_json(f"/repos/{repo.owner}/{repo.name}/commits/{quote(branch, safe='')}", token)
        if not isinstance(commit, dict):
            raise RepositoryProviderError(RepositoryProviderErrorCode.UNKNOWN, "Malformed GitHub response")
        return str(commit.get("sha", ""))

    async def walk_tree(
        self, token: str, repo: RepoRef, branch: str, subpath: str | None = None
    ) -> AsyncIterator[TreeEntry]:
        tree_response = await self._request_json(f"/repos/{repo.owner}/{repo.name}/git/trees/{quote(branch, safe='')}?recursive=1", token)
        if not isinstance(tree_response, dict):
            raise RepositoryProviderError(RepositoryProviderErrorCode.UNKNOWN, "Malformed GitHub response")

        tree = tree_response.get("tree", [])
        if not isinstance(tree, list):
            return

        prefix = _normalize_subpath(subpath)
        for entry in tree:
            if not isinstance(entry, dict):
                continue

            path = str(entry.get("path", ""))
            if not path:
                continue
            if prefix and not path.startswith(prefix):
                continue

            entry_type = _map_tree_type(str(entry.get("type", "")))
            if not entry_type:
                continue

            yield TreeEntry(
                path=path,
                type=entry_type,
                sha=str(entry.get("sha", "")),
                size_bytes=int(entry.get("size", 0)),
            )

    async def read_file(self, token: str, repo: RepoRef, branch: str, path: str) -> ReadFileResult:
        encoded_path = "/".join(quote(part, safe="") for part in path.strip("/").split("/") if part)
        file_response = await self._request_json(
            f"/repos/{repo.owner}/{repo.name}/contents/{encoded_path}?ref={quote(branch, safe='')}",
            token,
        )
        if not isinstance(file_response, dict):
            raise RepositoryProviderError(RepositoryProviderErrorCode.UNKNOWN, "Malformed GitHub response")

        content = str(file_response.get("content", "")).replace("\n", "")
        return ReadFileResult(
            content_base64=content,
            sha=str(file_response.get("sha", "")),
            size_bytes=int(file_response.get("size", 0)),
        )

    async def _request_json(self, path_with_query: str, token: str) -> Any:
        headers = {
            "Authorization": f"Bearer {token}",
            "Accept": "application/vnd.github+json",
            "X-GitHub-Api-Version": "2022-11-28",
        }

        client = self._get_or_create_client()
        return await _request_with_client(client, path_with_query, headers)


async def _request_with_client(client: httpx.AsyncClient, path_with_query: str, headers: dict[str, str]) -> Any:
    try:
        response = await client.get(path_with_query, headers=headers)
    except httpx.HTTPError as exc:
        raise RepositoryProviderError(RepositoryProviderErrorCode.NETWORK, "Failed to reach GitHub API") from exc

    if response.is_success:
        return response.json()

    body = response.text
    if response.status_code == 401:
        raise RepositoryProviderError(RepositoryProviderErrorCode.UNAUTHORIZED, "Unauthorized GitHub token", 401)
    if response.status_code == 404:
        raise RepositoryProviderError(RepositoryProviderErrorCode.NOT_FOUND, "GitHub resource not found", 404)
    if response.status_code == 429 or _is_rate_limited(response.status_code, body, response.headers):
        raise RepositoryProviderError(RepositoryProviderErrorCode.RATE_LIMITED, "GitHub API rate limit exceeded", response.status_code)
    raise RepositoryProviderError(
        RepositoryProviderErrorCode.UNKNOWN,
        f"Unexpected GitHub API error: {response.status_code}",
        response.status_code,
    )


def _is_rate_limited(status_code: int, body: str, headers: httpx.Headers) -> bool:
    if status_code != 403:
        return False
    if headers.get("x-ratelimit-remaining") == "0":
        return True
    return "rate limit" in body.lower()


def _normalize_subpath(subpath: str | None) -> str:
    if not subpath:
        return ""
    trimmed = subpath.strip("/")
    if not trimmed:
        return ""
    return f"{trimmed}/"


def _map_tree_type(raw: str) -> str | None:
    if raw == "blob":
        return "file"
    if raw == "tree":
        return "dir"
    if raw == "symlink":
        return "symlink"
    if raw == "commit":
        return "submodule"
    return None
